tools/webidl_binder.py

```diff
@@ -649,13 +649,13 @@ def make_call_args(i):
         js_call_args = ', '.join(['%s%s' % (('(ptrdiff_t)' if sig[j] in interfaces else '') + take_addr_if_nonpointer(raw[j]), args[j]) for j in range(i)])
 
         js_impl_methods.append(r'''  %s %s(%s) %s {
-    %sEM_ASM_%s({
+    %s (%s) EM_ASM_%s({
       var self = Module['getCache'](Module['%s'])[$0];
       if (!self.hasOwnProperty('%s')) throw 'a JSImplementation must implement all functions, you forgot %s::%s.';
       %sself['%s'](%s)%s;
     }, (ptrdiff_t)this%s);
   }''' % (c_return_type, func_name, dec_args, maybe_const,
-          basic_return, 'INT' if c_return_type not in C_FLOATS else 'DOUBLE',
+          basic_return, c_return_type, 'INT' if c_return_type not in C_FLOATS else 'DOUBLE',
```

Comment:
We have `EM_ASM_PTR` now that we should probably use for pointer returns.

I wonder if use that then we don't need the above cast?

Maybe create a new `em_asm_macro` variable here that can hold the name of the macro to use?

Reply:
This is a 2.5 years old PR. I can investigate with your hint and perhaps offer an improved version. I know very little about the project internals, however.

Reply:
I'm concerned that I'm getting test failures with the fixed code. Assuming they worked before, that likely means there are a bunch of tests that validate the existing wrong behavior...

Reply:
I refactored it with your suggestion, but the type cast cannot be avoided - the C++ compiler still raises an error for incompatible types. I incorporated the use of EM_ASM_PTR instead of EM_ASM_INT and extracted into a variable as you suggested.